### PR TITLE
C++ header gen: Made the enums into macros that users can redefine

### PR DESCRIPTION
**Pull request by TurkeyMan**
This is an attempt at the macro solution for the enums that we've talked about.

### CODE REVIEW

**Comment by TurkeyMan** [comment]
@ibuclaw How much does this upset you? I can't imagine any more simple and flexible ways to go with this.

File: src/dmd/dtoh.d [hunk 1]
```diff
@@ -235,6 +235,25 @@ extern(C++) void genCppHdrFiles(ref Modules ms)
     buf.writestring("#if !defined(_d_real)\n");
     buf.writestring("# define _d_real long double\n");
     buf.writestring("#endif\n");
+    buf.writeByte('\n');
```
Comment (thewilsonator) [inline src/dmd/dtoh.d]: Can you do this as one large q-string please.

Comment (thewilsonator) [reply src/dmd/dtoh.d]: Everything from the `#pragma once` to the end of this block of diff.

File: src/dmd/dtoh.d [hunk 2]
```diff
@@ -321,6 +347,39 @@ public:
...
+        }
+        return EnumKind.Other;
+    }
+
+    private void writeEnumTypeName(AST.Type type)
+    {
+        if (type.ty == AST.Tarray)
+        {
+            AST.TypeArray arr = cast(AST.TypeArray)type;
+            if (arr.next.ty == AST.Tchar)
```
Comment (thewilsonator) [inline src/dmd/dtoh.d]: Do we care about wstring or dstring enums? Its fine to leave that for another PR.

Comment (TurkeyMan) [reply src/dmd/dtoh.d]: One step ahead of you... press F5.

File: src/dmd/dtoh.d (hunk 1 above)
Comment (TurkeyMan) [reply src/dmd/dtoh.d]: I'm about to carve that whole block up and only emit fragments for the macros when they actually appear in the output.

Comment (thewilsonator) [reply src/dmd/dtoh.d]: OK.

File: src/dmd/dtoh.d [hunk 3]
```diff
@@ -235,6 +235,25 @@ extern(C++) void genCppHdrFiles(ref Modules ms)
...
+    buf.writestring("# define END_ENUM_NUMERIC(type, name, upper, lower) };\n");
+    buf.writestring("# define END_ENUM_TYPE(type, name, upper, lower) };\n");
+    buf.writestring("# define END_ANON_ENUM() };\n");
+    buf.writestring("# define END_ANON_ENUM_NUMERIC(type) };\n");
+    buf.writestring("# define ENUM_KEY_NUMERIC(type, name, value, enumName, upper, lower, abbrev) name = value,\n");
+    buf.writestring("# define ENUM_KEY_TYPE(type, name, value, enumName, upper, lower, abbrev) static type const name = value;\n");
+    buf.writestring("# define ANON_ENUM_KEY_NUMERIC(type, name, value) name = value,\n");
+    buf.writestring("# define ENUM_CONSTANT_NUMERIC(type, name, value) enum : type { name = value };\n");
+    buf.writestring("# define ENUM_CONSTANT(type, name, value) static type const name = value;\n");
+    buf.writestring("#endif\n");
```
Comment (WalterBright) [inline src/dmd/dtoh.d]: You can write this as simply:
```
buf.writestring("
#if !defined(_D_ENUM)
# define _D_ENUM
...
");
```
in one go. It's a lot easier to read. D is this way because I never liked the C way of not being nice about multiline string literals.

File: src/dmd/dtoh.d [hunk 4]
```diff
@@ -321,6 +347,41 @@ public:
...
+                break;
+        }
+        return EnumKind.Other;
+    }
+
+    private void writeEnumTypeName(AST.Type type)
+    {
+        if (type.ty == AST.Tarray)
+        {
+            AST.TypeArray arr = cast(AST.TypeArray)type;
```
Comment (WalterBright) [inline src/dmd/dtoh.d]: This is why D now has:
```
if (auto arr = type.isTypeArray())
   ...
```

File: src/dmd/dtoh.d [hunk 5]
```diff
@@ -966,100 +1034,165 @@ public:
...
-                hasBaseType = true;
-                break;
-            case AST.Tint32, AST.Tuns32, AST.Tenum: // by default, the base is an int
-                break;
-            default:
-                import dmd.root.string : toDString;
-                printf ("%s\n", ed.ident.toChars());
-                assert(0, ed.memtype.kind.toDString);
+            // check all keys have matching type
+            foreach (i, _m; *ed.members)
```
Comment (WalterBright) [inline src/dmd/dtoh.d]: `i` is unused

File: src/dmd/dtoh.d [hunk 6]
```diff
@@ -966,100 +1034,165 @@ public:
...
+            {
+                AST.EnumMember m = cast(AST.EnumMember)_m;
```
Comment (WalterBright) [inline src/dmd/dtoh.d]: Using:
```
auto m = _m.isEnumMember();
```
avoids a dangerous cast.

File: src/dmd/dtoh.d [hunk 7]
```diff
@@ -966,100 +1034,165 @@ public:
...
+            buf.writestring(")\n");
         }
 
-        if (!ed.members)
+        // emit constant for each member
+        foreach (i, _m; *ed.members)
         {
-            buf.writestring(";\n\n");
-            return;
+            AST.EnumMember m = cast(AST.EnumMember)_m;
```
Comment (WalterBright) [inline src/dmd/dtoh.d]: bad cast, bad, :-)

File: src/dmd/dtoh.d [hunk 8]
```diff
@@ -966,100 +1034,165 @@ public:
...
+
+            EnumKind memberKind = kind;
+            AST.Type memberType = type;
+            if (!memberType)
+            {
+                memberType = m.type;
+                memberKind = getEnumKind(memberType);
+            }
```
Comment (WalterBright) [inline src/dmd/dtoh.d]: This assigns things twice, and is a bit difficult to follow. How about the "single assignment" approach:
```
AST.Type memberType = type ? type : m.type;
EnumKind memberKind = type ? kind : getEnumKind(memberType);
```
? It's also 2 lines instead of 5.

Comment (WalterBright) [reply src/dmd/dtoh.d]: That also makes it possible for `memberKind` to be `const`, which is even better.

File: src/dmd/dtoh.d (hunk 3 above)
Comment (TurkeyMan) [reply src/dmd/dtoh.d]: I tried many permutations of this; it pollutes the output with extra newlines, improper indentations, and the newlines in the output are dependent on which OS was used to compile DMD (`"\n"` vs `"\r\n"`).

File: src/dmd/dtoh.d (hunk 4 above)
Comment (TurkeyMan) [reply src/dmd/dtoh.d]: Thanks!

Comment (TurkeyMan) [reply src/dmd/dtoh.d]: `error : no property isTypeArray for type dmd.mtype.Type, did you mean dmd.mtype.Type.isTypeAArray?`

File: src/dmd/dtoh.d (hunk 6 above)
Comment (TurkeyMan) [reply src/dmd/dtoh.d]: Okay. But how could it be possible that `members` in `EnumDeclaration` are not `EnumMember`s

File: src/dmd/dtoh.d (hunk 7 above)
Comment (TurkeyMan) [reply src/dmd/dtoh.d]: Are other kinds of members possible in an `EnumDeclaration`?

File: test/dshell/extra-files/cpp_header_gen/app.cpp [hunk 9]
```diff
@@ -1,4 +1,9 @@
 // This should be solved by the header generator
+
+#define BEGIN_ENUM(name, upper, lower) enum name {
+#define ENUM_KEY(type, name, value, enumName, upper, lower, abbrev) upper##name = value,
+#define END_ENUM(name, upper, lower) };
```
Comment (TurkeyMan) [inline test/dshell/extra-files/cpp_header_gen/app.cpp]: @ibuclaw DMD FE wants enum's like this right?

**Comment by TurkeyMan** [comment]
So, I think this is the best way forward... but maybe there's disagreement?
What are we doing here?

**Review by thewilsonator — APPROVED** [review]

File: test/dshell/extra-files/cpp_header_gen/app.cpp (hunk 9 above)
Comment (ibuclaw) [reply test/dshell/extra-files/cpp_header_gen/app.cpp]: Where the basetype isn't an int, a typedef is needed too.

Comment (TurkeyMan) [reply test/dshell/extra-files/cpp_header_gen/app.cpp]: They emit a different macro which would be defined differently.
There is a few variations of the macro where C++ has hard language separations on how it is able to express enums.

**Comment by TurkeyMan** [comment]
FWIW, I actually don't like the names of the macros I wrote here, and I'd take suggestions to improve them...

**Comment by ibuclaw** [comment]
I'm sure we'll remove them and re-add them all over again before next release goes out. ;-)
